utils/src/main/kotlin/storage/LocalFileStorage.kt

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (C) 2019 HERE Europe B.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package com.here.ort.utils.storage
+
+import com.here.ort.utils.log
+import com.here.ort.utils.safeMkdirs
+
+import java.io.File
+import java.io.InputStream
+
+/**
+ * A [FileStorage] that stores file in a [directory] of the local file system. The [read] and [write] operations are
+ * [Synchronized].
+ */
+class LocalFileStorage(
+    /**
+     * The directory used to store files in.
+     */
+    val directory: File,
+
+    /**
+     * If true, try to create the [directory] if it does not exist.
+     */
+    createDirectory: Boolean = true
+) : FileStorage {
+    init {
+        if (!directory.exists() && createDirectory) {
+            log.debug { "Creating directory '${directory.invariantSeparatorsPath}' for local file storage." }
+            directory.safeMkdirs()
+        }
+
+        require(directory.isDirectory) {
+            "Cannot use storage directory '${directory.invariantSeparatorsPath}' because it is not a directory."
+        }
+
+        if (createDirectory) {
+            directory.safeMkdirs()
+        } else {
+            require(directory.isDirectory) {
+                "Cannot use storage directory '${directory.invariantSeparatorsPath}' because it is not a directory."
+            }
+        }
+    }
+
+    @Synchronized
+    override fun read(path: String): InputStream {
+        val file = directory.resolve(path)
+
+        require(file.canonicalFile.startsWith(directory.canonicalFile)) {
```

Comment:
What is the purpose of this check? We already resolve `path` relative to `directory`, and below we already check that `file` really is a file.

Reply:
The purpose is to prevent files being read or written outside of the storage directory by using relative paths starting with "..".

Reply:
I'm not 100% sure but to me making this `isInsideDir` check makes sense.
Looking at [1] I think this catches an additional edge case by adding File.separator to the end of `directory.canonicalFile`. Does it make sense to add it? consider:
```
directory = /a/bc
file=/a/bcd/f
```

[1] https://stackoverflow.com/a/23030314

Reply:
This case is already handled by using `directory.resolve(path)` which adds the separator if required.

Reply:
I don't understand how `directory.resolve(path)` is related to the issue I just illustrated.
The problem is that the string passed to `startsWith` does not have a trailing `File.SeparatorChar` also not in some cases a `canonicalFile` strips it.

```
fun main()
{
    val directory = File("/home/user/")
    println(directory.canonicalFile)
}
```
outputs: `/home/user`

Reply:
This is not `String.startsWith` but Kotlin's `File.startsWith` which works on the components of the path, so trailing slashes do not matter. See: https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.io/java.io.-file/starts-with.html

Reply:
oh my bad, thanks for explaining!